### PR TITLE
Add cosine similarity metric

### LINKED ISSUE #3202: Add cosine similarity metric
## 🚀 Feature
It would be nice to add cosine similarity as a Metric.
Cosine similarity between feature vectors is often used in representation learning.

### PR DESCRIPTION
**Pull request by kzkadc**
Fixes #3202

Description:

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

### CODE REVIEW

File: ignite/contrib/metrics/cosine_similarity.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred = output[0].flatten(start_dim=1).detach()
+        y = output[1].flatten(start_dim=1).detach()
+        cos_similarities = torch.nn.functional.cosine_similarity(y_pred, y, dim=1, eps=self.eps)
+        self._sum_of_cos_similarities += torch.sum(cos_similarities).to(self._device)
+        self._num_examples += y.shape[0]
+
+    @sync_all_reduce("_sum_of_cos_similarities", "_num_examples")
+    def compute(self) -> Union[float, torch.Tensor]:
```
Comment (vfdev-5) [inline ignite/contrib/metrics/cosine_similarity.py:94]: Maybe, just float as the output type ?
```suggestion
    def compute(self) -> float:
```

**Review by vfdev-5 — COMMENTED** [review]
Thanks a lot for the PR @kzkadc !
I think we can put this implementation to `ignite.metrics` instead of contrib metrics as it does not have any 3rd party dependencies.

**Comment by vfdev-5** [comment]
@kzkadc thanks for the update. Let's put this implementation to `ignite.metrics` instead of contrib metrics

**Comment by kzkadc** [comment]
@vfdev-5 Thank you for your review! I moved it to `igite.metrics`.
I will also try to write a test.

File: ignite/metrics/cosine_similarity.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    @reinit__is_reduced
+    def reset(self) -> None:
+        self._sum_of_cos_similarities = torch.tensor(0.0, device=self._device)
+        self._num_examples = 0
+
+    @reinit__is_reduced
+    def update(self, output: Sequence[torch.Tensor]) -> None:
+        y_pred = output[0].flatten(start_dim=1).detach()
+        y = output[1].flatten(start_dim=1).detach()
+        cos_similarities = torch.nn.functional.cosine_similarity(y_pred, y, dim=1, eps=self.eps)
```
Comment (vfdev-5) [inline ignite/metrics/cosine_similarity.py:89]:
```suggestion
        cos_similarities = torch.nn.functional.cosine_similarity(
            y_pred, y, dim=1, eps=self.eps
        )
```
or maybe we can import `from torch.nn import functional as F` and do
```python
     cos_similarities = F.cosine_similarity(y_pred, y, dim=1, eps=self.eps)
```

**Comment by vfdev-5** [comment]
@kzkadc if auto code-formatting GH action is bothering you, you can disable it from your fork. https://github.com/pytorch/ignite/actions/workflows/code-style.yml

**Review by vfdev-5 — APPROVED** [review]
Thanks a lot for the PR @kzkadc !

**Comment by vfdev-5** [comment]
Hi @kzkadc , sorry, I was too quick to merge this PR which actually has failing xla job: https://github.com/pytorch/ignite/actions/runs/8351321508/job/22859463184
Could you please fix it in a follow-up PR?

**Comment by kzkadc** [comment]
@vfdev-5 sorry for my mistake. I fixed it in https://github.com/pytorch/ignite/pull/3213.
